### PR TITLE
fixes #19659 - capsule sync via bulk action

### PR DESCRIPTION
**Pull request by bbuckingham**
Moving the capsule sync to a bulk action.
This will ensure that if one capsule is down, that
the remaining syncs continue.

### CODE REVIEW

**Comment by bbuckingham** [comment]
[test]

**Comment by jlsherrill** [comment]
[test]

File: app/lib/actions/katello/capsule_content/sync.rb [hunk 1]
```diff
@@ -6,18 +6,30 @@ def resource_locks
...
+        end
+
         def humanized_name
-          _("Synchronize capsule content")
+          _("Synchronize smart proxy")
+        end
+
+        def humanized_input
+          ["'#{input['smart_proxy']['name']}'"] + super
         end
```
Comment (johnpmitsch) [inline app/lib/actions/katello/capsule_content/sync.rb]: Showing the name of the proxy in the task :+1: :100:

**Comment by johnpmitsch** [comment]
Getting this error after a sync but not sure if its related to your PR
```
 | NameError: uninitialized constant ProxyAPI::PulpNode
 | /home/vagrant/katello/app/models/katello/concerns/smart_proxy_extensions.rb:73:in `update_puppet_path'
 | /home/vagrant/katello/app/models/katello/concerns/smart_proxy_extensions.rb:81:in `refresh_with_puppet_path'
 | /home/vagrant/foreman/app/controllers/api/v2/smart_proxies_controller.rb:60:in `refresh'
```

File: app/controllers/katello/api/v2/capsule_content_controller.rb [hunk 2]
```diff
@@ -51,7 +51,7 @@ def remove_lifecycle_environment
     param :environment_id, Integer, :desc => N_('Id of the environment to limit the synchronization on')
     def sync
       find_environment if params[:environment_id]
-      task = async_task(::Actions::Katello::CapsuleContent::Sync, capsule_content, :environment => @environment)
+      task = async_task(::Actions::Katello::CapsuleContent::Sync, capsule_content.capsule, :environment_id => @environment.try(:id))
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/capsule_content_controller.rb]: Why the change to environment_id? It seems like it's just used to find the environment below (same question goes for the other ids too)

File: app/lib/actions/katello/capsule_content/sync.rb [hunk 3]
```diff
@@ -6,18 +6,30 @@ def resource_locks
...
 
-          environment = options.fetch(:environment, nil)
-          repository = options.fetch(:repository, nil)
-          content_view = options.fetch(:content_view, nil)
+          environment_id = options.fetch(:environment_id, nil)
+          environment = ::Katello::KTEnvironment.find(environment_id) unless environment_id.nil?
+          repository_id = options.fetch(:repository_id, nil)
+          repository = ::Katello::Repository.find(repository_id) unless repository_id.nil?
+          content_view_id = options.fetch(:content_view_id, nil)
+          content_view = ::Katello::ContentView.find(content_view_id) unless content_view_id.nil?
```
Comment (johnpmitsch) [inline app/lib/actions/katello/capsule_content/sync.rb]: This is more readable and I think would work the same
```
content_view = ::Katello::ContentView.find(content_view_id) if content_view_id
```

**Comment by jlsherrill** [comment]
@johnpmitsch i've seen that before, try restarting your dev env

File: test/actions/katello/repository_test.rb [hunk 4]
```diff
@@ -444,7 +444,8 @@ class CapsuleGenerateAndSyncTest < TestBase
 
     it 'plans' do
       plan_action(action, repository)
-      assert_action_planed_with(action, ::Actions::Katello::CapsuleContent::Sync, capsule_content, :repository => repository)
+      assert_action_planned_with(action, ::Actions::BulkAction, ::Actions::Katello::CapsuleContent::Sync,
+                                 [capsule_content.capsule], :repository_id => repository.id)
     end
```
Comment (johnpmitsch) [inline test/actions/katello/repository_test.rb]: How hard would it be to test two proxies syncing using bulk actions? That is some new functionality added by this PR and it would be nice to have the test coverage.

**Comment by johnpmitsch** [comment]
When I shut down http on one proxy, the other one kept syncing, :+1: here

File: app/controllers/katello/api/v2/capsule_content_controller.rb (hunk 2 above)
Comment (bbuckingham) [inline app/controllers/katello/api/v2/capsule_content_controller.rb]: With BulkAction, attempting to pass a hash containing AR resources resulted in plan phase errors similar to below; however, passing in a hash of IDs worked fine.
```
undefined method `to_hash' for #<Katello::ContentView:0x007f70248aa480>
```

I haven't found an alternative approach for passing in the resources, aside from passing in an array of arguments vs a hash.  If you have suggestions, please do let me know.  In most of Katello's usage of BulkAction, it does pass an array of arguments; however, I leaned away from it here, since we invoke the action differently depending upon scenario (e.g. cv + env vs repo).

**Comment by bbuckingham** [comment]
@johnpmitsch, Thanks for the feedback.  I pushed several updates based upon the comments.  I haven't found a good alternative yet to passing IDs vs resources for the BulkAction.  We can revisit if folks have other suggestions.

**Comment by bbuckingham** [comment]
Thanks @johnpmitsch!  Created http://projects.theforeman.org/issues/20323 for the ui enhancement.
